src/lib/operators/jit_operator/operators/jit_aggregate.cpp

```diff
@@ -0,0 +1,312 @@
+      break;
+    case AggregateFunction::Sum:
+    case AggregateFunction::Max:
+    case AggregateFunction::Min:
+      Assert(value.data_type() != DataType::String && value.data_type() != DataType::Null,
+             "Invalid data type for aggregate function.");
+      // The data type and nullability of the aggregate depend on the input value.
+      _aggregate_columns.push_back(
+          JitAggregateColumn{column_name, column_position, function, value,
+                             JitHashmapValue(value.data_type(), value.is_nullable(), _num_hashmap_values++)});
+      break;
+    case AggregateFunction::Avg:
+      Assert(value.data_type() != DataType::String && value.data_type() != DataType::Null,
+             "Invalid data type for aggregate function.");
+      // Average aggregates are computed by first computing two aggregates: a SUM and a COUNT
+      _aggregate_columns.push_back(
+          JitAggregateColumn{column_name, column_position, function, value,
+                             JitHashmapValue(value.data_type(), value.is_nullable(), _num_hashmap_values++),
+                             JitHashmapValue(DataType::Long, false, _num_hashmap_values++)});
+      break;
+    case AggregateFunction::CountDistinct:
+      Fail("Not supported");
+  }
+}
+
+void JitAggregate::add_groupby_column(const std::string& column_name, const JitTupleValue& value) {
+  auto column_position = _aggregate_columns.size() + _groupby_columns.size();
+  _groupby_columns.push_back(
+      JitGroupByColumn{column_name, column_position, value,
+                       JitHashmapValue(value.data_type(), value.is_nullable(), _num_hashmap_values++)});
+}
+
+const std::vector<JitAggregateColumn> JitAggregate::aggregate_columns() const { return _aggregate_columns; }
+
+const std::vector<JitGroupByColumn> JitAggregate::groupby_columns() const { return _groupby_columns; }
+
+void JitAggregate::_consume(JitRuntimeContext& context) const {
+  // We use index-based for loops in this function, since the LLVM optimizer is not able to properly unroll range-based
+  // loops, and we need the unrolling for proper specialization.
+
+  const auto num_groupby_columns = _groupby_columns.size();
+  const auto num_aggregate_columns = _aggregate_columns.size();
+
+  // Step 1: Compute hash value of the input tuple
+  uint64_t hash_value = 0;
+
+  // Compute a hash for each groupby column and combine the resulting hashes.
+  for (uint32_t i = 0; i < num_groupby_columns; ++i) {
+    hash_value = (hash_value << 5) ^ jit_hash(_groupby_columns[i].tuple_value, context);
+  }
+
+  // Step 2: Look up the rows with this hash in the hashmap.
+  auto& hash_bucket = context.hashmap.indices[hash_value];
+
+  bool found_match = false;
+  uint64_t row_index;
+  // Iterate over each row that produces this hash. Unless there is a hash collision, there should only be at most 1
+  // entry for each hash.
+  for (const auto& index : hash_bucket) {
```

Comment:
You said you have to use index-based loops in the comment?

Reply:
Added a comment.